src/freenet/support/math/MersenneTwister.java

```diff
@@ -78,4 +85,71 @@ public synchronized void setSeed(byte[] seed) {
 	/** {@inheritDoc} */
 	@Override protected synchronized int next(int bits) { return super.next(bits); }
 
+	/**
+	 * Creates a new random number generator using a random seed from the provided source.
+	 * Note that this is significantly slower than the unsynchronized (hence not thread-safe)
+	 * variant that can be obtained from {@link #createUnsynchronized}.
+	 **/
+	public static MersenneTwister createSynchronized(byte[] seed) {
+		return new MersenneTwister(Fields.bytesToInts(seed));
+	}
+
+	/**
+	 * Creates a new random number generator using a byte array seed.
+	 * The returned generator is not synchronized and is therefore NOT thread-safe.
+	 **/
+	public static MersenneTwister createUnsynchronized(byte[] seed) {
+		return new Unsynchronized(seed);
+	}
+
+	private static final class Unsynchronized extends MersenneTwister {
+		private Unsynchronized(byte[] seed) {
+			super(Fields.bytesToInts(seed));
+		}
+
+		@Override
+		protected int next(int bits) {
+			return unsynchronizedNext(bits);
+		}
+
+		@Override
+		public void setSeed(int seed) {
+			unsynchronizedSetSeed(seed);
+		}
+
+		@Override
+		public void setSeed(int[] seed) {
+			unsynchronizedSetSeed(seed);
+		}
+
+		@Override
+		public void setSeed(long seed) {
+			unsynchronizedSetSeed(seed);
+		}
+
+		@Override
+		public void setSeed(byte[] seed) {
+			unsynchronizedSetSeed(seed);
+		}
+	}
+
+	final int unsynchronizedNext(int bits) {
```

Comment:
I don’t like that we’re adding non-`private` methods here, but oh well, I can’t think of anything else that’s not a mess, either…

Reply:
Yep, this was the best thing I could think of without breaking the class hierarchy or changing the default behaviour. At least this does not pollute or break the *public* API.